contributors/design-proposals/pod-priority-api.md

```diff
@@ -0,0 +1,187 @@
+
+## Overview 
+
+This design doc introduces the concept of priorities for pods in Kubernetes and
+how the priority impacts scheduling and preemption of pods when the cluster
+runs out of resources. A pod can specify a priority at the creation time. The
+priority must be one of the valid values and there is a total order on the
+values. The priority of a pod is independent of its workload type. The priority
+is global and not specific to a particular namespace.
+
+## Detailed Design 
+
+### Effect of priority on scheduling 
+
+One could generally expect a pod with higher priority has a higher chance of
+getting scheduled than the same pod with lower priority. However, there are
+many other parameters that affect scheduling decisions. So, a high priority pod
+may or may not be scheduled sooner than lower priority pods. The details of
+what determines the order at which pods are scheduled are beyond the scope of
+this document.
+
+### Effect of priority on preemption 
+
+Generally, lower priority pods are more likely to get preempted by higher
+priority pods when cluster is out of resources. In such a case, scheduler may
+decide to preempt lower priority pods to release enough resources for higher
+priority pending pods. As mentioned before, there are many other parameters
+that affect scheduling decisions, such as affinity and anti-affinity. If
+scheduler determines that a high priority pod cannot be scheduled even if lower
+priority pods are preempted, it will not preempt lower priority pods. Scheduler
+may have other restrictions on preempting pods, for example, it may refuse to
+preempt a pod if PodDisruptionBudget is violated. The details of scheduling and
+preemption decisions are beyond the scope of this document.
+
+### Priority in PodSpec 
+
+Pods may have priority in their pod spec. PodSpec has field called
+"PriorityName" which is specified by user. PodStatus has a corresponding field
+called "Priority". User-specified priority (PriorityName) is a string and all
+of the valid priority names are defined by a system wide mapping that maps each
+string to an integer. The PriorityName specified in a pod spec must be found in
+this map or the pod creation request will be rejected. If PriorityName is
+empty, "tier3" (see below) is assumed. I am open to other suggestions about
+what the default priority should be. Once the PriorityName is resolved to an
+integer, it is placed in "Priority" field of PodStatus.
+
+
+```
+type PodSpec struct {
+  ...
+  PriorityName string
+}
+
+type PodStatus struct {
+  ...
+  Priority int  // Populated by Admission Controller.
+}
+```
+
+### Priority names 
+
+A Kubernetes cluster has four predefined (built-in) priority names. The top
+priority is reserved for the system pods and its value is the largest 32-bit
+integer. The following list is an example of how the priorities and their
+values may look like. I am open to suggestions for the names and values of the
```

Comment:
It might be good to give an example of what kinds of workloads might map to each priority level. For example in decreasing order of priority

* Kubernetes system daemons (per-node like fluentd, and cluster-level like Heapster)
* critical user infrastructure (e.g. storage servers, monitoring system like Prometheus, etc.)
* components that are in the user-facing request serving path and must be able to scale up arbitrarily in response to load spikes (web servers, middleware, etc.)
* important interruptible workloads that need strong-ish guarantee of schedulability and of not being interrupted
* less important interruptible workloads that need a less strong guarantee of schedulability and of not being interrupted
* best effort